---
title: "Detecting and correcting batch effects with BEclear"
author: 
    - name: David Johann-Nepomuk Peter Rasp
      affiliation:
        Center for Bioinformatics, Saarland University, Saarbruecken, Germany
      email: David.J.Rasp@gmail.com
package: BEclear
abstract: >
    We show in this tutorial how to use the BEclear [@Akulenko2016] package to detect and correct
    batch effects in methylation data. package. Even though BEclear was developed for
    the use on methylation data, it can also be used to find and correct batch 
    effects in other kinds of data.
    The central method of BEclear is based on Latent Factor Models [@Candes2009], which can in 
    theory be used on every matrix containing real numbers to predict missing values.
date: "`r Sys.Date()`"
output: BiocStyle::html_document
bibliography: ../inst/REFERENCES.bib
vignette: >
  %\VignetteIndexEntry{BEclear tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction


# Installation

`r BiocStyle::Biocpkg("BEclear")` is available on Bioconductor. To install it 
you can therefore use the `r BiocStyle::Biocpkg("BiocManager")`:

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("BEclear")
```

Otherwise you can also install `r BiocStyle::Githubpkg("David-J-R/BEclear")` from 
from its Github repository by the following command:

```{r, eval=FALSE}
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
install_github("David-J-R/MoDentify")
```

We however recommend installing it through Bioconductor, as this takes care for you in 
installing the dependencies for you and furthermore you can refer to the release of 
Bioconductor, when using our package, which enables you to reproduce the exact 
conditions of your run.

When done with the installation you can simply load the package by typing:

```{r}
library(BEclear)
```

# Data sources

```{r data}
data("BEclearData")
```

# Detecting of batch effects

```{r detection, cache=TRUE}
batchEffect <- calcBatchEffects(data = ex.data, samples = ex.samples, 
                                adjusted = TRUE, method = "fdr")
mdifs <- batchEffect$med
pvals <- batchEffect$pval
summary <- calcSummary(medians=mdifs, pvalues=pvals)
score <- calcScore(ex.data, ex.samples, summary, dir=getwd())

makeBoxplot(ex.data, ex.samples, score, bySamples=TRUE,
        col="standard", main="Example data", xlab="Batch",
        ylab="Beta value", scoreCol=TRUE)
```

# Imputation of missing values

For the imputation of missing values we use a gradient descent method developed 
by @Koren2009. In this section we will describe you our implementation of this
method and how to use it.

## Theoretical background

We assume that our complete data matrix \(D_{ij}\) can be described by the effects of
a matrix \(L_i\), which represents the effect of the features (genes in our case)
and a matrix \(R_j\) describing the effect of the samples in the following way:

\begin{equation}
D_{ij} = L_{i}^{T} \times R_{j}
(\#eq:assumption)
\end{equation}

The method can either be run on the complete data set or the data set can be 
divided into blocks on which the method is applied.
This divison into blocks allows for parallelisation of the method, which can be 
useful to speed up the process. We have found that a blocksize of 60x60 works 
well.[@Akulenko2016]

\begin{equation}
  errorMatrix_{ij} = Block_{ij} - L_{i}^{T} \times R_{j}
  (\#eq:errormatrix)
\end{equation}

We try to minimalise the following loss function through a gradient descent:

\begin{equation}
  min_{L, R}  \sum_{ij \in K}(errorMatrix_{ij}^2) + \lambda \times
  (\left\lVert L_{i}\right\rVert_{F}^{2} + 
  \left\lVert R_{j}\right\rVert_{F}^{2} )
  (\#eq:loss)
\end{equation}
Where \(K\) is the set of tuples \((i,j)\) for which the value is present. 
And \(\lambda\) is the penalty coefficient, which controls how restrictive the 
selection of variables should be. The default of \(\lambda\) is 1.

For the first epoch the matrices \(L_i\) and \(R_j\) are filled with random values
generated by the `rnorm` function from the `r BiocStyle::Rpackage("stats")` 
package and the initial loss and error matrix are calculated.

Then for each epoch the following is done:
    
* \(L_i\) and \(R_j\) are modified proportional to \(\gamma\) by the following 
calculation:

\begin{equation}
L_i = L_i + \gamma \times (errorMatrix_{ij} \times R_j - \lambda \times L_i)
(\#eq:Lmod)
\end{equation}

\begin{equation}
R_j = R_j + \gamma \times (errorMatrix_{ij} \times L_i - \lambda \times R_j)
(\#eq:Rmod)
\end{equation}

## Usage

First you have to set the found batch effects values to NAs You can do this
by using the `clearBEgenes` function:

```{r clearBE, cache=TRUE}
cleared.data <- clearBEgenes(ex.data, ex.samples, summary)
```
In case you're using `r BiocStyle::Biocpkg("BEclear")` not for correcting batch
effects, but just for the data imputation, you would have to set the values you
want to impute to NA, if they not already are. 

For the data imputation you use the `imputeMissingData` function:

```{r imputation, cache=TRUE}
corrected.data <- imputeMissingData(cleared.data, rowBlockSize=60, 
                                    colBlockSize=60,epochs=50, 
                                    outputFormat="", dir=getwd())
```

If you set rowBlockSize and colBlockSize to 0 the matrix won't be divided into 
block and the gradient descent will be applied to the matrix as a whole.


```{r boxplot}
makeBoxplot(corrected.data, ex.samples, score, bySamples=TRUE,
        col="standard", main="Corrected example data",
        xlab="Batch", ylab="Beta value", scoreCol=FALSE)
```

# Overall correction

Besides the individual methods BEclear also offers an overall method, which 
executes all the described previous step in a call. It also applies some 
preprocessing to your data set

```{r correction, cache=TRUE}
cleared.data <- correctBatchEffect(data=ex.data, samples=ex.samples)
```
# sessionInfo()

Here is the output of `sessionInfo()` on the system on which this document 
was compiled running pandoc `r rmarkdown::pandoc_version()`:

```{r, echo=FALSE}
sessionInfo()
```

# References